### PR TITLE
Fix share_memory_ compile

### LINKED ISSUE #166623: torch.compile fails with share_memory_() operation: '_share_fd_: only available on CPU'
### 🐛 Describe the bug

🐛 Bug Description
torch.compile fails when the model contains share_memory_() operations during the compilation phase. The error occurs because share_memory_() cannot be executed on FakeTensor objects used by Dynamo for graph tracing.
code:
'''
import os
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
import matplotlib.pyplot as plt
from torch.utils.data import Dataset, DataLoader, random_split

class AutoRegressiveDataset(Dataset):

    def __init__(self, data, window_size=20, horizon=5, stride=1):
        self.data = data
        self.window_size = window_size
        self.horizon = horizon
        self.stride = stride
        self.length = (len(data) - window_size - horizon) // stride + 1

    def __len__(self):
        return self.length

    def __getitem__(self, idx):
        start = idx * self.stride
        end = start + self.window_size
        target_end = end + self.horizon
        window = self.data[start:target_end]
        (min_val, max_val) = (window.min(), window.max())
        normalized = 2 * (window - min_val) / (max_val - min_val) - 1
        x = normalized[:self.window_size]
        y = normalized[self.window_size:]
        return (torch.FloatTensor(x).unsqueeze(1), torch.FloatTensor(y).unsqueeze(1))

class AutoRegressiveRNN(nn.Module):

    def __init__(self, input_size=1, hidden_size=64, output_size=1, num_layers=2, horizon=5):
        super().__init__()
        self.hidden_size = hidden_size
        self.num_layers = num_layers
        self.horizon = horizon
        self.rnn = nn.LSTM(input_size=input_size, hidden_size=hidden_size, num_layers=num_layers, batch_first=True)
        self.fc = nn.Linear(hidden_size, output_size)

    def init_hidden(self, batch_size, device):
self.hidden_size).to(device),torch.zeros(self.num_layers, batch_size, self.hidden_size).to(device))
        return (torch.zeros(self.num_layers, batch_size, self.hidden_size).share_memory_().to(device), torch.zeros(self.num_layers, batch_size, self.hidden_size).to(device))

    def forward(self, x, pred_steps=None):
        batch_size = x.size(0)
        hidden = self.init_hidden(batch_size, x.device)
        if pred_steps is None:
            (rnn_out, _) = self.rnn(x, hidden)
            outputs = self.fc(rnn_out)
            return outputs[:, -self.horizon:, :]
        else:
            predictions = []
            current_input = x
            (h, c) = hidden
            for _ in range(pred_steps):
                (rnn_out, (h, c)) = self.rnn(current_input, (h, c))
                pred = self.fc(rnn_out[:, -1, :])
                predictions.append(pred)
                current_input = torch.cat([current_input[:, 1:, :], pred.unsqueeze(1)], dim=1)
            return torch.stack(predictions).squeeze(1)

def get_default_model():
    input_size = 1
    hidden_size = 64
    output_size = 1
    num_layers = 2
    horizon = 5
    model = AutoRegressiveRNN(input_size=input_size, hidden_size=hidden_size, output_size=output_size, num_layers=num_layers, horizon=horizon)
    return model

def get_sample_inputs():
    batch_size = 4
    seq_len = 20
    input_size = 1
    x = torch.randn(batch_size, seq_len, input_size)
    return (x,)

def main():
    model = get_default_model()
    model.eval()
    inputs = get_sample_inputs()

    with torch.no_grad():
        output_normal = model(*inputs)
    print('Normal inference successful')
    print(f'Output shape: {output_normal.shape}')

    try:
        compiled_model = torch.compile(model)
        compiled_model.eval()
        with torch.no_grad():
            output_compiled = compiled_model(*inputs)
        print('Compiled inference successful')
        print(f'Compiled output shape: {output_compiled.shape}')

        diff = torch.abs(output_normal - output_compiled).max().item()
        print(f'Output difference: {diff:.6f}')

    except Exception as exception:
        print("Compilation failed:", exception)

    print(f'Model parameters: {sum((p.numel() for p in model.parameters()))}')

if __name__ == '__main__':
    main()
'''
output:
'''
Normal inference successful
Output shape: torch.Size([4, 5, 1])

Compilation failed: Dynamo failed to run FX node with fake tensors: call_method share_memory_(*(FakeTensor(..., size=(2, 4, 64)),), **{}): got RuntimeError('_share_fd_: only available on CPU')

from user code:
line 50, in forward
    hidden = self.init_hidden(batch_size, x.device)
line 46, in init_hidden
    return (torch.zeros(self.num_layers, batch_size, self.hidden_size).share_memory_().to(device), torch.zeros(self.num_layers, batch_size, self.hidden_size).to(device))

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"

Model parameters: 50497
'''

### Versions

PyTorch version: 2.7.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.5.0-6ubuntu2) 9.5.0
Clang version: Could not collect
CMake version: version 4.0.3
Libc version: glibc-2.39

Python version: 3.9.7 (default, Jul 16 2025, 16:34:47)  [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-29-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU
Nvidia driver version: 580.65.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 39 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 32
On-line CPU(s) list: 0-31
Vendor ID: GenuineIntel
Model name: Intel(R) Core(TM) i9-14900HX
CPU family: 6
Model: 183
Thread(s) per core: 2
Core(s) per socket: 24
Socket(s): 1
Stepping: 1
CPU(s) scaling MHz: 31%
CPU max MHz: 5800.0000
CPU min MHz: 800.0000
BogoMIPS: 4838.40
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization: VT-x
L1d cache: 896 KiB (24 instances)
L1i cache: 1.3 MiB (24 instances)
L2 cache: 32 MiB (12 instances)
L3 cache: 36 MiB (1 instance)
NUMA node(s): 1
NUMA node0 CPU(s): 0-31
Vulnerability Gather data sampling: Not affected
Vulnerability Ghostwrite: Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] nvidia-cublas-cu11==11.10.3.66
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu11==11.7.101
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu11==11.7.99
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu11==11.7.99
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu11==8.5.0.96
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu11==10.2.10.91
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu11==11.4.0.1
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu11==11.7.4.91
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu11==2.14.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu11==11.7.91
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.7.1
[pip3] torchdiffeq==0.2.5
[pip3] torchvision==0.22.1
[pip3] triton==3.3.1
[conda] Could not collect

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @Lucaskabela

### PR DESCRIPTION
**Pull request by parsshar-RH**
Fixes #166623

### Summary:
Fixes share_memory_ in compile mode

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @mlazos

### CODE REVIEW

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
Seems like a good fix! The CI failure comes from an unexpected success, so we should update the test so that it is not expected to fail - try deleting

`test/dynamo_expected_failures/TestMultiprocessing.test_inherit_tensor`

Then feel free to resubmit and ping me :)

**Comment by Lucaskabela** [comment]
Hm yeah I suggested removing this test because the signals showed it had passed (https://github.com/pytorch/pytorch/actions/runs/20453728203/job/58772360831)

So now that it is failing, seems that this is a race condition/flaky test in the current impl; so a graph break and fall back to eager seems good here.  Lets try this

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
See feedback; lets try graph breaking here for now

**Review by Lucaskabela — APPROVED** [review]
Thanks for the changes - lgtm
